python/lsst/resources/gs.py

```diff
@@ -0,0 +1,279 @@
+
+from ._resourcePath import ResourcePath
+
+if TYPE_CHECKING:
+    from .utils import TransactionProtocol
+
+log = logging.getLogger(__name__)
+
+
+_client = None
+"""Cached client connection."""
+
+
+def _get_client() -> storage.Client:
+    global _client
+    if _client is None:
+        _client = storage.Client()
+    return _client
+
+
+class GSResourcePath(ResourcePath):
+    """Access Google Cloud Storage resources."""
+
+    _bucket: Optional[storage.Bucket] = None
+    _blob: Optional[storage.Blob] = None
+    _client: Optional[storage.Client] = None
+
+    @property
+    def client(self) -> storage.Client:
+        return _get_client()
+
+    @property
+    def bucket(self) -> storage.Bucket:
+        if self._bucket is None:
+            self._bucket = self.client.bucket(self.netloc)
+        return self._bucket
+
+    @property
+    def blob(self) -> storage.Blob:
+        if self._blob is None:
+            self._blob = self.bucket.blob(self.relativeToPathRoot)
+        return self._blob
+
+    def exists(self) -> bool:
+        if self.is_root:
+            return self.bucket.exists()
+        return self.blob.exists()
+
+    def size(self) -> int:
+        if self.dirLike:
+            return 0
+        # The first time this is called we need to sync from the remote.
+        # Should we track when this needs to be called again?
+        self.blob.reload()
+        size = self.blob.size
+        if size is None:
+            raise FileNotFoundError(f"Resource {self} does not exist")
+        return size
+
+    def remove(self) -> None:
+        try:
+            self.blob.delete()
+        except exceptions.NotFound as e:
+            raise FileNotFoundError(f"No such resource: {self}") from e
+
+    def read(self, size: int = -1) -> bytes:
+        if size < 0:
+            start = None
+            end = None
+        else:
+            start = 0
+            end = size
+        try:
+            with time_this(log, msg="Read from %s", args=(self,)):
+                body = self.blob.download_as_bytes(start=start, end=end)
+        except exceptions.NotFound as e:
+            raise FileNotFoundError(f"No such resource: {self}") from e
+        return body
+
+    def write(self, data: bytes, overwrite: bool = True) -> None:
+        if not overwrite:
+            if self.exists():
+                raise FileExistsError(f"Remote resource {self} exists and overwrite has been disabled")
+        with time_this(log, msg="Write to %s", args=(self,)):
+            self.blob.upload_from_string(data)
+
+    def mkdir(self) -> None:
+        if not self.bucket.exists():
+            raise ValueError(f"Bucket {self.netloc} does not exists for {self}!")
+
+        if not self.dirLike:
+            raise NotADirectoryError(f"Can not create a 'directory' for a file-like UEI {self}")
+
+        if self.is_root:
+            # The root must already exist.
+            return
+
+        # Should this method do anything at all?
```

Comment:
Yes. S3 does it and this matches that behavior. I'm not really sure what the best approach is in an object store. Make `mkdir()` a no-op and then always say a "directory" exists because in theory it could?